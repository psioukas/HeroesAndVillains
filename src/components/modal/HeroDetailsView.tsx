import DeleteIcon from '@mui/icons-material/Delete';
import { Box, BoxProps, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import { observer } from 'mobx-react-lite';
import { useEffect } from 'react';
import Store from '../../store';
import HeroApiRequests from '../../utils/HeroApiRequests';
import Button from '../Button';
import HeroAvatar from '../hero/HeroAvatar';

const StyledHeroDetailsView = styled(Box)<BoxProps>(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    [theme.breakpoints.up('sm')]: {
        gap: theme.spacing(2),
    },
    [theme.breakpoints.down('sm')]: {
        height: '100%',
        '& .btnDelete': {
            position: 'absolute',
            bottom: theme.spacing(4.5),
            left: 0,
            right: 0,
        },
    },
}));

const HeroDetailsView = () => {
    const hero = Store.selectedHero;
    useEffect(() => {
        if (hero && !window.location.pathname.includes(hero.id)) {
            window.history.pushState(undefined, '', `${window.location.origin}/details/${hero.id}`);
        }
        return () => {
            window.history.pushState(undefined, '', `${window.location.origin}`);
        };
    }, [hero]);

    const handleDeleteHero = async () => {
        if (!hero) return;
        // const successfull: boolean = await HeroApiRequests.deleteHero(hero.id);
        // if (successfull) {
        Store.notification.show('Please wait deleting hero', 'info');
        setTimeout(() => {
        Store.deleteHero(hero);
            setTimeout(() => {
                Store.notification.show('Hero deleted successfully', 'success');
            }, 100);
        }, 1100);
    };

    return hero ? (
        <StyledHeroDetailsView data-testid='hero-details-view'>
            <HeroAvatar
                src={hero.avatarUrl}
                alt="hero avatar"
                width={94}
                height={94}
                style={{ alignSelf: 'center' }}
                mt={Store.isMobile ? 10 : 0}
            />

            <Typography
                alignSelf={'center'}
                fontWeight={'bold'}
                variant="body1"
                mt={Store.isMobile ? 10 : 0}
                mb={Store.isMobile ? 0.5 : 0}
            >
                {hero.fullName}
            </Typography>
            <Typography alignSelf={'center'} variant="body2" mb={Store.isMobile ? 8 : 0}>
                {hero.type.name}
            </Typography>
            <Typography alignSelf={'center'} variant="body2">
                {hero.description}
            </Typography>
            <Button
                variant={'text'}
                textColor={'#bb4d5f'}
                startIcon={<DeleteIcon />}
                onClick={handleDeleteHero}
                className={'btnDelete'}
            >
                Delete hero
            </Button>
        </StyledHeroDetailsView>
    ) : null;
};

export default observer(HeroDetailsView);
