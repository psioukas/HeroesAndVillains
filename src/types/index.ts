export interface IHeroType {
    id: string;
    name: string;
}

export interface IHero {
    id: string;
    fullName: string;
    avatarUrl: string;
    description: string;
    type: IHeroType;
}

export type CreateHeroType = Omit<IHero, 'type' | 'id'> & {
    typeId: string;
};

export interface IHeroesListResponse {
    data: IHero[];
    totalCount: number;
}
