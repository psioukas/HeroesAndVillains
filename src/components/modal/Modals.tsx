import CloseIcon from '@mui/icons-material/Close';
import {Box, BoxProps, IconButton, Modal, Paper, PaperProps, SvgIconProps, Typography,} from '@mui/material';
import {styled} from '@mui/material/styles';
import {observer} from 'mobx-react-lite';
import Store from '../../store';
import AddHeroView from './AddHeroView';
import HeroDetailsView from './HeroDetailsView';
import AddHeroTypeView from "./AddHeroTypeView.tsx";
import UpdateHeroView from "./UpdateHeroView.tsx";

const StyledModalBody = styled(Paper)<PaperProps>(({theme}) => ({
    position: 'absolute',
    paddingBlock: theme.spacing(6),
    background: theme.palette.background.default,
    outline: 'none',
    [theme.breakpoints.up('sm')]: {
        width: theme.spacing(90),
        left: '50%',
        top: '50%',
        transform: 'translate(-50%,-50%)',
        paddingInline: theme.spacing(9),
    },
    [theme.breakpoints.down('sm')]: {
        height: '100%',
        width: '100%',
        left: 0,
        top: 0,
        paddingInline: theme.spacing(4.5),
    },
}));

const StyledModalHeader = styled(Box)<BoxProps>(({theme}) => ({
    display: 'flex',
    alignItems: 'center',
    height: theme.spacing(6.25),
}));

const StyledModalContent = styled(Box)<BoxProps>(({theme}) => ({
    width: '100%',
    height: `calc(100% - ${theme.spacing(6.25)})`,
}));

const StyledIconButton = styled(IconButton)<SvgIconProps>(() => ({
    color: '#7f7f7f',
    marginLeft: 'auto',
    '&:hover ': {
        color: '#4d4d4d',
    },
    '&: active': {
        color: '#272727',
    },
}));
const Modals = () => {
    return (
        <Modal
            open={Store.modals.isAnyModalOpen}
            onClose={Store.modals.closeOpenModal}
            sx={{
                width: Store.isMobile ? '100%' : 'auto',
            }}
        >
            <StyledModalBody>
                <StyledModalHeader>
                    {Store.modals.activeModalTitle && (
                        <Typography variant={'h6'}>{Store.modals.activeModalTitle}</Typography>
                    )}
                    <StyledIconButton
                        data-testid='close-button'
                        aria-label="close modal"
                        disableFocusRipple
                        disableTouchRipple
                        onClick={Store.modals.closeOpenModal}
                    >
                        <CloseIcon fontSize={'medium'}/>
                    </StyledIconButton>
                </StyledModalHeader>
                <StyledModalContent>
                    {Store.modals.addHero.visible && <AddHeroView/>}
                    {Store.modals.addHeroType.visible && <AddHeroTypeView/>}
                    {Store.modals.updateHero.visible && <UpdateHeroView/>}
                    {Store.modals.heroDetails.visible && <HeroDetailsView/>}
                </StyledModalContent>
            </StyledModalBody>
        </Modal>
    );
};
export default observer(Modals);
