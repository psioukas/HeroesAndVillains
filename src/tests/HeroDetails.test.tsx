import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import HeroDetailsView from '../components/modal/HeroDetailsView';
import Modals from '../components/modal/Modals';
import Store from '../store';
import { IHero } from '../types';
window.HTMLElement.prototype.scrollIntoView = jest.fn();

const hero: IHero = {
    id: 'whatever-id',
    avatarUrl: 'http://www.google.com/images/wtv',
    fullName: 'Full Metal',
    type: {
        id: 'id-human',
        name: 'human',
    },
    description: 'dangerous',
};

const mockedFetch = jest.fn(() =>
    Promise.resolve({
        json: () => Promise.resolve({}),
    }),
) as jest.Mock;

global.fetch = mockedFetch;

beforeEach(() => {
    Store.addHero(hero);
    Store.showHeroDetails(hero.id);
});
afterEach(() => {
    Store.clearSelectedHero();
    Store.deleteHero(hero);
});

test('Expect Snapshot to match', () => {
    const domTree = renderer.create(<HeroDetailsView />).toJSON();
    expect(domTree).toMatchSnapshot();
});
test('Renders Hero Details', () => {
    render(<Modals />);
    const heroDetailsView = screen.getByTestId('hero-details-view');
    expect(heroDetailsView).toBeInTheDocument();
});

test('Click Modal close button, should close Hero Details view', () => {
    render(<Modals />);
    const heroDetailsView = screen.getByTestId('hero-details-view');
    expect(heroDetailsView).toBeInTheDocument();

    const button = screen.getByTestId('close-button');
    expect(button).toBeInTheDocument();

    userEvent.click(button);

    const activeModalTitle = Store.modals.activeModalTitle;
    const heroDetailsVisible = Store.modals.heroDetails.visible;

    expect(activeModalTitle).toBe('');
    expect(heroDetailsVisible).toBe(false);
});
test('Check hero details are in the view', () => {
    render(<HeroDetailsView />);
    const fullNameText = screen.getByText(hero.fullName);
    const descriptionText = screen.getByText(hero.description);
    expect(fullNameText).toBeInTheDocument();
    expect(descriptionText).toBeInTheDocument();
});
