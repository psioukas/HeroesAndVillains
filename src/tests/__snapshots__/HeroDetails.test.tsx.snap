// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect Snapshot to match 1`] = `
<div
  className="MuiBox-root css-10jbkkq"
  data-testid="hero-details-view"
>
  <div
    className="MuiBox-root css-3wml6g"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    <img
      alt="hero avatar"
      height={94}
      src="http://www.google.com/images/wtv"
      width={94}
    />
  </div>
  <p
    className="MuiTypography-root MuiTypography-body1 css-rfvo0n-MuiTypography-root"
  >
    Full Metal
  </p>
  <p
    className="MuiTypography-root MuiTypography-body2 css-16ucqvx-MuiTypography-root"
  >
    human
  </p>
  <p
    className="MuiTypography-root MuiTypography-body2 css-147ip0x-MuiTypography-root"
  >
    dangerous
  </p>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation btnDelete css-176uzx2-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
    Delete hero
  </button>
</div>
`;
